debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/transforms/MongoArrayConverterTest.java

```diff
@@ -349,4 +390,102 @@ public void shouldCreateStructForHeterogenousDocumentInArray() throws Exception
+    }
+
+    @Test
+    public void shouldCreateSchemaForNestedDocumentForDocumentEncoding() {
+        final MongoDataConverter converter = new MongoDataConverter(ArrayEncoding.DOCUMENT);
+        final BsonDocument val = BsonDocument.parse(NESTED_DOCUMENT);
+        Map<String, Map<Object, BsonType>> entry = converter.parseBsonDocument(val);
+        converter.buildSchema(entry, builder);
+
+        final Schema finalSchema = builder.build();
+
+        final Schema expectedSchema = SchemaBuilder.struct().name("array")
+                .field("pipeline", SchemaBuilder.struct().name("array.pipeline").optional()
+                        .field("_0", SchemaBuilder.struct().name("array.pipeline._0").optional()
+                                .field("stageId", Schema.OPTIONAL_INT32_SCHEMA)
+                                .field("componentList", SchemaBuilder.struct().name("array.pipeline._0.componentList").optional()
+                                        .field("_0", SchemaBuilder.struct().name("array.pipeline._0.componentList._0").optional()
+                                                .field("componentId", Schema.OPTIONAL_INT32_SCHEMA)
+                                                .field("action", Schema.OPTIONAL_STRING_SCHEMA)
+                                                .build())
+                                        .build())
+                                .build())
+                        .build())
+                .build();
+
+        assertThat(finalSchema).isEqualTo(expectedSchema);
+
+        final Struct struct = new Struct(finalSchema);
+        for (Map.Entry<String, BsonValue> bsonValueEntry : val.entrySet()) {
+            converter.buildStruct(bsonValueEntry, finalSchema, struct);
+        }
+
+        final String expectedStruct = "Struct{pipeline=Struct{_0=Struct{stageId=1,componentList=Struct{_0=Struct{componentId=1,action=deploy}}}}}";
+        assertThat(struct.toString()).isEqualTo(expectedStruct);
+    }
+
+    @Test
+    public void shouldCreateSchemaForNestedSubDocumentForArrayEncoding() {
+        final MongoDataConverter converter = new MongoDataConverter(ArrayEncoding.ARRAY);
+        final BsonDocument val = BsonDocument.parse(NESTED_SUB_DOCUMENT);
+        Map<String, Map<Object, BsonType>> entry = converter.parseBsonDocument(val);
+        converter.buildSchema(entry, builder);
+
+        final Schema finalSchema = builder.build();
+        final Struct struct = new Struct(finalSchema);
+        for (Map.Entry<String, BsonValue> bsonValueEntry : val.entrySet()) {
+            converter.buildStruct(bsonValueEntry, finalSchema, struct);
+        }
+
+        assertThat(struct.toString()).isEqualTo(
+                "Struct{pipeline=[Struct{stageId=1,componentList=[Struct{componentId=1,action=deploy}, Struct{componentId=2,action=deploy}]}, Struct{stageId=2,componentList=[Struct{componentId=3,action=deploy}, Struct{componentId=4,action=deploy}]}]}");
```

Comment:
All such assertions can use formatted multile String that will just remove endlines before comparison.